src/Interpreters/TraceCollector.cpp

```diff
@@ -56,9 +56,17 @@ TraceCollector::~TraceCollector()
   */
 void TraceCollector::stop()
 {
-    WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
-    writeChar(true, out);
-    out.next();
+    try
+    {
+        WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
+        writeChar(true, out);
+        out.next();
+    }
+    catch (...)
+    {
+        tryLogCurrentException("TraceCollector");
```

Comment:
not good because the thread will never join if the character is not sent right?

Reply:
yep

Reply:
I think the character will not be sent only if pipe is broken, and if so, it will be detected in run() and thread will end and joined.

Reply:
yes, but we can get OOM exception in WriteBuffer so we won't send the character and pipe will be still alive

Reply:
Well, it's mostly two generals' problem and it's not solvable - even if we add some synchronization besides flag communicated through pipe - it's the same problem. The only thing I can offer is to detach and abandon thread in case of exception - but in this case we will have the possibility of stuck running thread... maybe additionally close the pipe so to force thread to finish on its own because of exception on read?